2.0/m_checkbans.cpp

```diff
@@ -0,0 +1,274 @@
+/*
+ * InspIRCd -- Internet Relay Chat Daemon
+ *
+ *   Copyright (C) 2017 genius3000 <genius3000@g3k.solutions>
+ *
+ * This file is a module for InspIRCd.  InspIRCd is free software: you can
+ * redistribute it and/or modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* $ModAuthor: genius3000 */
+/* $ModAuthorMail: genius3000@g3k.solutions */
+/* $ModDesc: Adds commands /checkbans, /testban, and /whyban */
+/* $ModDepends: core 2.0 */
+
+/* Helpop Lines for the CUSER section
+ * Find: '<helpop key="cuser" value="User Commands
+ * Place 'TESTBAN  WHYBAN    CHECKBANS' after 'SSLINFO'
+ * Re-space as needed to match the current columns
+ * Find: '<helpop key="sslinfo" ...'
+ * Place just above that line:
+<helpop key="testban" value="/TESTBAN <channel> <mask>
+
+Test a channel ban mask against the users currently in the channel.">
+
+<helpop key="whyban" value="/WHYBAN <channel> [user]
+
+Get a list of bans and exceptions that match you (or the given user)
+on the specified channel.">
+
+<helpop key="checkbans" value="/CHECKBANS <channel>
+
+Get a list of bans and exceptions that match current users on the channel.">
+
+ */
+
+
+#include "inspircd.h"
+#include "u_listmode.h"
+
+/* Common function */
+bool CanCheck(Channel* c, User* u)
+{
+	if (u->HasPrivPermission("channels/auspex"))
+		return true;
+
+	ModeHandler* mh = ServerInstance->Modes->FindMode('b', MODETYPE_CHANNEL);
+	return (mh->GetLevelRequired() <= c->GetPrefixValue(u));
+}
+
+/* Handle /CHECKBANS */
+class CommandCheckBans : public Command
+{
+ public:
+	CommandCheckBans(Module* Creator) : Command(Creator, "CHECKBANS", 1, 1)
+	{
+		this->syntax = "<channel>";
+	}
+
+	CmdResult Handle(const std::vector<std::string>& parameters, User* user)
+	{
+		Channel* c = ServerInstance->FindChan(parameters[0]);
+		if (!c)
+		{
+			user->WriteNumeric(ERR_NOSUCHNICK, "%s %s :No such channel", user->nick.c_str(), parameters[0].c_str());
+			return CMD_FAILURE;
+		}
+
+		/* Only allow checking for matching users if you have access to the ban list. */
+		if (!CanCheck(c, user))
+		{
+			user->WriteNumeric(ERR_CHANOPRIVSNEEDED, "%s %s :You do not have access to the ban list.",
+				user->nick.c_str(), c->name.c_str());
+			return CMD_FAILURE;
+		}
+
+		/* Check for channel mode 'e', ban exceptions */
+		ListModeBase* ex = (ListModeBase*)ServerInstance->Modes->FindMode('e', MODETYPE_CHANNEL);
```

Comment:
You should check the mode name here before casting it as 'e' might be bound to another mode.